### PR TITLE
    KAFKA-14265: Prefix ACLs may shadow other prefix ACLs

### PR DESCRIPTION
**Pull request by cmccabe**
Prefix ACLs may shadow other prefix ACLs. Consider the case where we have prefix ACLs for foobar, fooa, and f. If we were matching a resource named "foobar", we'd start scanning at the foobar ACL, hit the fooa ACL, and stop – missing the f ACL.

To fix this, we should re-scan for ACLs at the first divergence point (in this case, f) whenever we hit a mismatch of this kind.

### CODE REVIEW

File: metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java [hunk 1]
```diff
@@ -394,35 +399,61 @@ private MatchingAclRule findAclRule(
...
-                if (acl.patternType() == LITERAL && !resourceName.equals(acl.resourceName())) {
+            int matchesUpTo = matchesUpTo(resourceName, acl.resourceName());
+            if (matchesUpTo == acl.resourceName().length()) {
+                if (acl.patternType() == LITERAL && matchesUpTo != resourceName.length()) {
                     // This is a literal ACL whose name is a prefix of the resource name, but
                     // which doesn't match it exactly. We should skip over this ACL, but keep
                     // scanning in case there are any relevant PREFIX ACLs.
                     continue;
                 }
+
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java:441]: nit: extra new line

**Review by hachikuji — COMMENTED** [review]
Good find. The fix LGTM, but we probably need some test cases.

**Comment by cmccabe** [comment]
I added two tests. One unit test and one integration test. I verified that both fail without the fix. The integration tests succeeds with AclAuthorizer prior to the StandardAuthorizer fix.

**Review by hachikuji — APPROVED** [review]
LGTM

**Comment by hachikuji** [comment]
The build is failing due to the newly added test: `kafka.api.AuthorizerIntegrationTest.testPrefixAcls(String).quorum=zk`

**Comment by ijuma** [comment]
Looks like we didn't squash this PR before merging. Let's avoid that next time please.

**Pull request merged**